### PR TITLE
Generate Unit Tests for Benchmarks

### LINKED ISSUE #5222: Benchmarks should generate unit tests
A runtime benchmark is basically the same as a unit test:

* It sets up storage.
* Runs a bunch of functions which cannot fail, else it returns an error.
* Ultimately executes a final extrinsic which must return `Ok(())`.

The `benchmark!` macro should wrap the benchmark logic into simple unit tests which confirm that the benchmarks will not fail when actually being executed for benchmarking.

Ideally, it would run some similar loop over the components just like a real benchmark, maybe with a limited number of steps and no repeats to reduce the time it takes to execute.

This also means, as extrinsics change, or benchmarks are written, that these tests will execute as a part of the CI process for pull requests.

### PR DESCRIPTION
**Pull request by shawntabrizi**
Closes #5222 

This PR introduces automatically generated unit tests for each runtime benchmark.

From the docs:

Test functions are automatically generated for each benchmark and are accessible to you when you run `cargo test`. All tests are named `test_benchmark_<benchmark_name>`, expect you to pass them the Runtime Trait, and run them in a test externalities environment. The test function runs your benchmark just like a regular benchmark, but only testing at the lowest and highest values for each component. The function will return `Ok(())` if the benchmarks return no errors.

You can construct benchmark tests like so:

```rust
#[test]
fn test_benchmarks() {
  new_test_ext().execute_with(|| {
    assert_ok!(test_benchmark_dummy::<Test>());
    assert_err!(test_benchmark_other_name::<Test>(), "Bad origin");
    assert_ok!(test_benchmark_sort_vector::<Test>());
    assert_err!(test_benchmark_broken_benchmark::<Test>(), "You forgot to sort!");
  });
}
```

### CODE REVIEW

File: frame/treasury/src/benchmarking.rs [hunk 1]
```diff
@@ -217,3 +217,25 @@ benchmarks! {
...
+	fn test_benchmarks() {
+		new_test_ext().execute_with(|| {
+			assert_ok!(test_benchmark_propose_spend::<Test>());
+			assert_ok!(test_benchmark_reject_proposal::<Test>());
+			assert_ok!(test_benchmark_approve_proposal::<Test>());
+			assert_ok!(test_benchmark_report_awesome::<Test>());
+			assert_ok!(test_benchmark_retract_tip::<Test>());
+			// assert_ok!(test_benchmark_tip_new::<Test>());
+			// assert_ok!(test_benchmark_tip::<Test>());
+			// assert_ok!(test_benchmark_close_tip::<Test>());
```
Comment (shawntabrizi) [inline frame/treasury/src/benchmarking.rs:237]: There is an issue where I am unable to modify the fixed TenToFourteen Contains struct.

This is needed to test these benchmarks, and I could use help on how to do that,

File: primitives/state-machine/src/in_memory_backend.rs [hunk 2]
```diff
@@ -364,6 +364,10 @@ impl<H: Hasher> Backend<H> for InMemory<H> where H::Out: Codec {
 	fn usage_info(&self) -> UsageInfo {
 		UsageInfo::empty()
 	}
+
+	fn wipe(&self) -> Result<(), Self::Error> {
+		Ok(())
+	}
```
Comment (shawntabrizi) [inline primitives/state-machine/src/in_memory_backend.rs:370]: This seems to wipe the in memory backend just like I want... but why? All I did was return `Ok(())`

File: frame/session/benchmarking/src/lib.rs [hunk 3]
```diff
@@ -42,16 +42,33 @@ benchmarks! {
...
+// 	use frame_support::assert_ok;
+
+// 	#[test]
+// 	fn test_benchmarks() {
+// 		new_test_ext().execute_with(|| {
+// 			assert_ok!(test_benchmark_set_keys::<Test>());
+// 			assert_ok!(test_benchmark_purge_keys::<Test>());
+// 		});
+// 	}
+// }
```
Comment (shawntabrizi) [inline frame/session/benchmarking/src/lib.rs:74]: For some reason, this code is unable to find `palle_session::mock`... why?

**Comment by shawntabrizi** [comment]
The last CI thing seems to be an issue with the VM

File: frame/session/benchmarking/src/lib.rs (hunk 3 above)
Comment (shawntabrizi) [reply frame/session/benchmarking/src/lib.rs]: You cannot access resources behind `#[cfg(test)]`, so I had to create a new mock for just the benchmarking setup.

File: frame/treasury/src/benchmarking.rs (hunk 1 above)
Comment (shawntabrizi) [reply frame/treasury/src/benchmarking.rs]: I used the refcell pattern here to create a mutable list of members which I then referenced in the `TenToFourteen` struct to get me the right information.

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -305,7 +305,7 @@ test-runtime-benchmarks:
       - $DEPLOY_TAG
   script:
     - cd bin/node/cli
-    - BUILD_DUMMY_WASM_BINARY=1 time cargo check --verbose --features runtime-benchmarks
+    - BUILD_DUMMY_WASM_BINARY=1 time cargo test --verbose --features runtime-benchmarks
```
Comment (shawntabrizi) [inline .gitlab-ci.yml:308]: Does it make sense to have my own CI which runs all tests again with this feature flag?

Or should I simply add my feature flag to the main testing?

File: primitives/state-machine/src/in_memory_backend.rs (hunk 2 above)
Comment (shawntabrizi) [reply primitives/state-machine/src/in_memory_backend.rs:370]: To be clear, everything works, I just want to know why...

File: .gitlab-ci.yml (hunk 4 above)
Comment (gavofyork) [reply .gitlab-ci.yml:308]: i guess it's probably fine to just add it to the main testing.

Comment (shawntabrizi) [reply .gitlab-ci.yml]: ahh, I cant do it actually because the main test works off of the virtual manifest, which does not accept feature flags :/

**Pull request merged**